File: creative-studio-ui/src/services/n8nService.ts
/**
 * n8n Management Service
 */

export interface n8nStatus {
  status: 'online' | 'offline';
  message: string;
}

export interface n8nWorkflow {
  id: string;
  name: string;
  active: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface n8nTemplate {
  name: string;
  filename: string;
  description?: string;
  tags?: string[];
}

class N8nService {
  private baseUrl: string;

  constructor() {
    this.baseUrl = (import.meta as unknown as { env: Record<string, string> }).env.VITE_API_URL || 'http://localhost:8080';
  }

  async getStatus(): Promise<n8nStatus> {
    try {
      const response = await fetch(`${this.baseUrl}/api/n8n/status`);
      if (!response.ok) throw new Error('Failed to fetch n8n status');
      return await response.json() as n8nStatus;
    } catch (_error) {
      // Background check - log as info only
      console.debug('n8n getStatus: Service not reachable (expected if backend is down)');
      return { status: 'offline', message: 'n8n not reachable' };
    }
  }

  async listWorkflows(): Promise<n8nWorkflow[]> {
    try {
      const response = await fetch(`${this.baseUrl}/api/n8n/workflows`);
      if (!response.ok) throw new Error('Failed to fetch n8n workflows');
      const data = await response.json() as { workflows: n8nWorkflow[] };
      return data.workflows || [];
    } catch (_error) {
      // Background check - log as info only
      console.debug('n8n listWorkflows: Service not reachable (expected if backend is down)');
      return [];
    }
  }

  async triggerWorkflow(webhookId: string, payload: Record<string, unknown>): Promise<unknown> {
    try {
      const response = await fetch(`${this.baseUrl}/api/n8n/trigger/${webhookId}`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(payload),
      });
      if (!response.ok) throw new Error('Failed to trigger n8n workflow');
      return await response.json();
    } catch (error) {
      console.error('n8n triggerWorkflow error:', error);
      throw error;
    }
  }

  async createWorkflow(name: string, nodes: unknown[], connections: Record<string, unknown>): Promise<unknown> {
    try {
      const response = await fetch(`${this.baseUrl}/api/n8n/workflows`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ name, nodes, connections }),
      });
      if (!response.ok) throw new Error('Failed to create n8n workflow');
      return await response.json();
    } catch (error) {
      console.error('n8n createWorkflow error:', error);
      throw error;
    }
  }

  async listTemplates(): Promise<n8nTemplate[]> {
    try {
      const response = await fetch(`${this.baseUrl}/api/n8n/templates`);
      if (!response.ok) throw new Error('Failed to fetch n8n templates');
      return await response.json() as n8nTemplate[];
    } catch (_error) {
      // Background check - log as info only
      console.debug('n8n listTemplates: Service not reachable (expected if backend is down)');
      return [];
    }
  }

  async importTemplate(filename: string): Promise<unknown> {
    try {
      const response = await fetch(`${this.baseUrl}/api/n8n/templates/import/${filename}`, {
        method: 'POST'
      });
      if (!response.ok) throw new Error('Failed to import n8n template');
      return await response.json();
    } catch (error) {
      console.error('n8n importTemplate error:', error);
      throw error;
    }
  }

  getN8nUrl(): string {
    return (import.meta as unknown as { env: Record<string, string> }).env.VITE_N8N_URL || 'http://192.168.1.47:5678/home/workflows';
  }
}

export const n8nService = new N8nService();
export default n8nService;

File: backend/n8n_service.py
import logging
import aiohttp
from typing import List, Dict, Any
from backend.config import settings

logger = logging.getLogger(__name__)


class N8nService:
    """
    Service for interacting with n8n instance on the local network.
    """

    def __init__(self):
        self.base_url = settings.N8N_BASE_URL.rstrip("/")
        self.api_key = settings.N8N_API_KEY
        self.webhook_url = settings.N8N_WEBHOOK_URL.rstrip("/")

    async def check_status(self) -> Dict[str, Any]:
        """Check if n8n is reachable."""
        try:
            async with aiohttp.ClientSession() as session:
                async with session.get(
                    f"{self.base_url}/healthz", timeout=5
                ) as response:
                    if response.status == 200:
                        return {"status": "online", "message": "n8n is reachable"}
                    else:
                        return {
                            "status": "offline",
                            "message": f"n8n returned status {response.status}",
                        }
        except Exception as e:
            logger.error(f"Error checking n8n status: {e}")
            return {"status": "offline", "message": str(e)}

    async def list_workflows(self) -> List[Dict[str, Any]]:
        """List workflows from n8n (requires API key)."""
        if not self.api_key:
            return []

        try:
            headers = {"X-N8N-API-KEY": self.api_key}
            async with aiohttp.ClientSession() as session:
                async with session.get(
                    f"{self.base_url}/api/v1/workflows", headers=headers
                ) as response:
                    if response.status == 200:
                        data = await response.json()
                        return data.get("data", [])
                    else:
                        logger.error(f"n8n API returned status {response.status}")
                        return []
        except Exception as e:
            logger.error(f"Error listing n8n workflows: {e}")
            return []

    async def trigger_workflow(
        self, webhook_id: str, payload: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Trigger an n8n workflow via webhook."""
        try:
            url = f"{self.webhook_url}/{webhook_id}"
            async with aiohttp.ClientSession() as session:
                async with session.post(url, json=payload) as response:
                    if response.status in [200, 201]:
                        return {"success": True, "data": await response.json()}
                    else:
                        return {"success": False, "error": f"Status {response.status}"}
        except Exception as e:
            logger.error(f"Error triggering n8n workflow {webhook_id}: {e}")
            return {"success": False, "error": str(e)}

    async def create_workflow(
        self, name: str, nodes: List[Dict[str, Any]], connections: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Create a new workflow in n8n."""
        if not self.api_key:
            return {"success": False, "error": "API key not configured"}

        try:
            url = f"{self.base_url}/api/v1/workflows"
            headers = {
                "X-N8N-API-KEY": self.api_key,
                "Content-Type": "application/json",
            }
            payload = {
                "name": name,
                "nodes": nodes,
                "connections": connections,
                "settings": {},
            }

            async with aiohttp.ClientSession() as session:
                async with session.post(url, headers=headers, json=payload) as response:
                    if response.status in [200, 201]:
                        return {"success": True, "data": await response.json()}
                    else:
                        error_text = await response.text()
                        return {
                            "success": False,
                            "error": f"Status {response.status}: {error_text}",
                        }
        except Exception as e:
            logger.error(f"Error creating n8n workflow: {e}")
            return {"success": False, "error": str(e)}


# Singleton instance
n8n_service = N8nService()


def get_n8n_service():
    return n8n_service
